src/Symfony/Component/PropertyAccess/PropertyAccessor.php

```diff
@@ -131,23 +132,66 @@ public function setValue(&$objectOrArray, $propertyPath, $value)
             self::IS_REF => true,
         ));
 
-        for ($i = count($propertyValues) - 1; $i >= 0; --$i) {
-            $objectOrArray = &$propertyValues[$i][self::VALUE];
+        try {
+            if (PHP_VERSION_ID < 70000) {
+                self::$previousErrorHandler = set_error_handler(array(__CLASS__, 'handleError'));
+            }
 
-            if ($overwrite) {
-                $property = $propertyPath->getElement($i);
-                //$singular = $propertyPath->singulars[$i];
-                $singular = null;
+            for ($i = count($propertyValues) - 1; $i >= 0; --$i) {
+                $objectOrArray = &$propertyValues[$i][self::VALUE];
 
-                if ($propertyPath->isIndex($i)) {
-                    $this->writeIndex($objectOrArray, $property, $value);
-                } else {
-                    $this->writeProperty($objectOrArray, $property, $singular, $value);
+                if ($overwrite) {
+                    $property = $propertyPath->getElement($i);
+                    //$singular = $propertyPath->singulars[$i];
+                    $singular = null;
+
+                    if ($propertyPath->isIndex($i)) {
+                        $this->writeIndex($objectOrArray, $property, $value);
+                    } else {
+                        $this->writeProperty($objectOrArray, $property, $singular, $value);
+                    }
                 }
+
+                $value = &$objectOrArray;
+                $overwrite = !$propertyValues[$i][self::IS_REF];
+            }
+        } catch (\TypeError $e) {
+            try {
+                self::throwUnexpectedTypeException($e->getMessage(), $e->getTrace(), 0);
+            } catch (UnexpectedTypeException $e) {
```

Comment:
throwing is done conditionally (see throwUnexpectedTypeException). Thus if the methods throws, $e changes, otherwise $e stays as the original TypeError. This is exactly the logic we need here.